Framework/Algorithms/src/BinaryOperation.cpp

```diff
@@ -492,7 +492,7 @@ void BinaryOperation::doSingleValue() {
 
   if (m_eout) {
     // ---- The output is an EventWorkspace ------
-    PARALLEL_FOR3(m_lhs, m_rhs, m_out)
+    PARALLEL_FOR_IF(Kernel::threadSafe(m_lhs.get(), m_rhs.get(), m_out.get()))
```

Comment:
One should be able to use any combination of pointers and references. Pointers are checked for  `nullptr`, while references must never be NULL.

I changed the functions that were generating warnings with GCC 6.  Some of the three parameter versions can probably also be changed.